### ISSUE TITLE
Document that the Start-Sleep cmdlet is misnamed

### ISSUE THREAD

**Issue by mklement0** [MicrosoftDocs/PowerShell-Docs#4474]
The `Start-Sleep` cmdlet is invariably _synchronous_ in nature: it blocks the invoking thread until the specified interval has elapsed.

By contrast, the approved-verb guidelines recommend that `Start` be used for _asynchronous_ tasks, as correctly in the name of the `Start-Process` cmdlet, for instance; it is the verb `Invoke` that is meant to convey _synchronous_ operations.

From https://docs.microsoft.com/en-us/powershell/developer/cmdlet/approved-verbs-for-windows-powershell-commands#invoke-vs-start (emphasis added):

> The **Invoke** verb is used to perform an operation that is generally a **synchronous** operation, such as running a command. The **Start** verb is used to begin an operation that is generally an **asynchronous** operation, such as starting a process.

Therefore, the proper name for this cmdlet is `Invoke-Sleep`.

Based on the discussion in https://github.com/PowerShell/PowerShell/issues/3990, the consensus  seems to be that it's not worth introducing an `Invoke-Sleep` alias or to rename the cmdlet and make `Start-Sleep` an alias.

However, to spare future readers potential confusion, please add a to the `Start-Sleep` help topic note to point out this inconsistency.

Version(s) of document impacted
------------------------------
- [x] Impacts 7 document
- [x] Impacts 6 document
- [x] Impacts 5.1 document
- [x] Impacts 5.0 document
- [x] Impacts 4.0 document
- [x] Impacts 3.0 document

**Comment by KirkMunro** [comment]
I don't find `Start-Sleep` to be inappropriately named at all, especially given that it invokes `System.Threading.Thread.Sleep`, which is used to temporarily suspect the execution of the current thread for a specified amount of time, and during that time other threads may get the CPU for execution. That tie-in to asynchronous execution is close enough to me to this definition, especially since it says "generally" in the details, so no change is necessary here in my opinion.

> The **Start** verb is used to begin an operation that is generally an asynchronous operation, such as starting a process.

@mklement0 While it may not appear to have anything to do with asynchronous operation, since PowerShell is typically used synchronously, since we have PSThreadJob as well as `S.M.A.PowerShell.InvokeAsync` and `S.M.A.PowerShell.StopAsync` as public methods now, there are more opportunities for asynchronous invocation of PowerShell in a single process than in the past. In my mind that makes the `Start` verb in `Start-Sleep` be acceptable and appropriate. Do you see the correlation I'm trying to make here?

If you still feel there is confusion that should be addressed, maybe the docs for `Start-Sleep` should identify the scenarios where it can be used, which include adding delays for certain tasks to complete where those tasks may return before the work has fully completed, or when you are working with PowerShell in multiple threads or runspaces and you want to sleep the current thread to allow another thread to have a chance to get access to the CPU.

That's just my opinion on the topic, the review committee may think otherwise.

**Comment by mklement0** [comment]
@KirkMunro:

I think what matters with respect to command naming is whether the command is synchronous or asynchronous _from the caller's perspective_, and from that perspective `Start-Sleep` is a misnomer.

> maybe the docs for Start-Sleep should identify the scenarios where it can be used,

👍

**Comment by KirkMunro** [comment]
Technically it also starts a timed sleep. I still feel the verb is very appropriate, much more than Invoke.

**Comment by mklement0** [comment]
It's not about what happens _technically_ - which may not even be on a user's radar.

From the caller's perspective, it's a _blocking_ (synchronous) call that doesn't return until the specified interval has elapsed: "sleep for a while, then continue with the next statement".

Let's look at the other standard cmdlets named `Start-*`:

~~~
Start-Job
Start-Process
Start-ThreadJob
Start-Transcript
~~~

What do they - appropriately - have in common?
They all _initiate_ a _background activity_ and (virtually) _instantly return control to the caller_.

**Comment by joeyaiello** [comment]
I think the key word in the original verb documentation is "generally". I don't think that's implied to mean "must be".

@mklement0 : Were you or anyone you know confused by this in the wild when experimenting with `Start-Sleep`. While not impossible, I'd be a little surprised if anyone who was playing with `Start-Sleep` wouldn't see that it's a synchronous sleep.

**Comment by mklement0** [comment]
@joeyaiello: If the consensus is that - except for yours truly - no one has been or ever will be confused by this inconsistency, no action is needed, and no further debate.

---

If, by contrast, we put pragmatism aside and take your argument seriously, in an effort to gain a better understanding of the rules, particularly with an eye toward future practice:

> I think the key word in the original verb documentation is "generally". 

My reading of "generally" is: "That's how it should be, _except if there's a good reason to make an exception_."

To me, it makes sense  to pick verbs for cmdlets  in manner that reflects  their "synchronicity" with respect to the behavior _experienced by the caller_. It's a simple rule, and to me the most straightforward interpretation.

If you disagree, please explain why.

If you agree, please explain what makes the purpose of `Start-Sleep` exceptional so as to warrant deviating from the recommended nomenclature.

If you feel there's sufficient uncertainty as to what the best choice is for a given cmdlet - along the lines of @KirkMunro's reasoning - the answer is obvious: _document_ the choice that _was_ made, to help those confused by the uncertainty.

**Comment by joeyaiello** [comment]
> My reading of "generally" is: "That's how it should be, except if there's a good reason to make an exception."

"Generally" could just as easily be a statement about a common side effect that occurs because of some other property. I'm at a loss for coming up with a good example, but here goes: I'd say that people that live near coasts "generally" like the ocean because they've grown up around it, but I wouldn't say that they need to have a reason not to like the ocean. It's just a common observation that one characteristic commonly results in another characteristic. 

> To me, it makes sense to pick verbs for cmdlets in manner that reflects their "synchronicity" with respect to the behavior experienced by the caller. It's a simple rule, and to me the most straightforward interpretation.

To me, the most relevant aspect of Start vs. Invoke is in their descriptions further down the page:

> Invoke: Performs an action, such as running a command or a method.
> Start: Initiates an operation. For example, the Start-Service cmdlet starts a service. This verb is paired with Stop.

Through this lens, as well as those of the dictionary definitions of the words, `Start-Sleep` just makes more sense to me. I don't invoke a sleep, it's not a .NET method, it's not a CIM method, it's not a native command, it's not a cmdlet. The runspace "starts sleeping" when I run the command. 

So I *guess* we could reword the "Invoke vs. Sleep" docs to say something along the lines of "the nature of invocations often results Invoke-* being used with synchronous operations", or something along those lines? But I think the reality is that people are going to use "Invoke" when they're calling into a thing and "Start" when they're kicking a long-running process.

**Comment by mklement0** [comment]
**tl;dr:**  

* I think that the existing [Approved Verbs for PowerShell Commands](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/developer/cmdlet/approved-verbs-for-windows-powershell-commands.md) topic is fine as-is.

* I personally think that _only_  the `Start-Sleep` topic needs amending, but I am aware that, so far, apparently, no one else deems that necessary. 
  * See below for why I still think that doing so is called for -  but I understand that continuing this debate may be of little interest to many.

----

Truthfully, I drowned in your ocean metaphor, if you will - I didn't get it.

> I don't invoke a sleep, it's not a .NET method, it's not a CIM method, it's not a native command, it's not a cmdlet.

Yes, you _do_ invoke a sleep _operation_:  Behind the scenes, you're invoking  the [`Thread.Sleep`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.thread.sleep) _method_.

However, the _specifics_ of that operation's _implementation_ are irrelevant - what matters is its "synchronicity".

>  The runspace "starts sleeping" when I run the command.

Yes, it _starts_ sleeping, but it _also waits for that sleep to complete_, and that's what matters in terms of nomenclature.

It's important to keep the following two aspects separate:

* How long it takes to complete a given operation.

* Whether the caller is _blocked_ while that operation is being performed, i.e., whether the operation is _synchronous_ from the perspective fo the caller.

An operation that takes a long to time to complete may call for _asynchronous_ operation - e.g., launching a background job - but it is the _method of invocation_ alone that determines _actual_ "synchronicity".

That is, you may opt to sleep for _days_ with `Start-Sleep`, but to you as the caller that still is a _syncrhonous_ operation, and therefore an `Invoke-*` operation.

**Comment by mklement0** [comment]
I just noticed that the same confusion affects the `build.psm1` module that is used to define build-related functions for building PowerShell itself (though a far smaller audience will be exposed to these functions):

All of the following functions are misnamed, because they operate _synchronously_:

```
Start-CrossGen
Start-NativeExecution
Start-PSBootstrap
Start-PSBuild
Start-PSPester
Start-PSxUnit
Start-ResGen
Start-TypeGen
Start-UnelevatedProcess
```

The only correctly named function is `Start-DevPowerShell`, which (at least on Windows) is asynchronous, because it opens a new window without blocking the calling shell.

Perhaps the misnaming here stems from the idea that the verb `Start` suggests a _lengthy operation_; however, the prescribed synchronous (`Invoke`) / asynchronous (`Start`) distinction is independent of how long a command takes to run.

**Comment by KirkMunro** [comment]
I like these names. They start processes. Sure, right now they wait for those processes, showing the output in the console, but they still start processes. Plus, if we get [this RFC](https://github.com/PowerShell/PowerShell-RFC/pull/193) accepted and implemented, we can make these async invocations that show output by default in an overlay/window depending on the host that they are invoked from.

Wrt discussions about commands using the Start verb inappropriately, we should just accept that Start is not only for asynchronous operations. Start is a natural verb for starting some process, task or job. Invoke is a natural verb for invoking a command. Neither of those should be specifically tied to async or sync invocations. That's what Start-Job, Start-ThreadJob and Start-Process, along with the `&` and hopefully `&!` and `~&` and `~&!` (see [this RFC for the last two operators](https://github.com/PowerShell/PowerShell-RFC/pull/205)) to control sync va async invocations in PowerShell.

We just need to update the docs that describe when you would use these verbs. It's that simple.

**Comment by mklement0** [comment]
> Neither of those should be specifically tied to async or sync invocations. 

They _are_ in the current guidelines, and thankfully so.

And the majority of the standard cmdlets observe this distinction, thankfully so.

It's a useful distinction: the verb `Start` tells me:

* An operation is _initiated_, but _isn't (necessarily) complete_ when the call returns.
* I'm free to do other things while the operation continues _elsewhere_, and if I care about when / with what output the operation completes, I need to do extra work.

> but they still start processes. 

The verb "start" loses all definition with this vague interpretation.
Anything your code calls "starts" something.
There would be no merit in having separate `Start` and `Invoke` verbs, and avoiding that kind of redundancy / ambiguity is what standardized verbs is all about.

> We just need to update the docs that describe when you would use these verbs. It's that simple.

The docs already do that, and, to my mind, sufficiently.

`Start-Job`, `Start-ThreadJob` and `Start-Process` are all appropriate named.

That `Start-Process` offers _opting into_ synchronous behavior is immaterial, because a cmdlet should be named for its _default_ behavior.

It is similarly immaterial that features _external_ to a given command - such as the post-positional `&` operator - can change the synchrony of a call.

To be clear: I think what your RFCs propose would make for useful additions to the language, but they are unrelated to our conversation here.

**Comment by KirkMunro** [comment]
No.

Async invocation is important is when you know up front that you have an asynchronous process that can be started, and perhaps stopped, suspended, resumed, waited for, etc., in which case there are verbs for that: `Start`, `Stop`, `Suspend`, `Resume`, `Wait`. In that scenario, those verbs are important to know about and they should be used for commands that manage asynchronous processing. e.g. the set of `Job`/`ThreadJob` cmdlets, or the set of `Process` cmdlets.

This is the only case where synchronous vs asynchronous execution should be used as a decision point when choosing a verb for a command -- when you know a command is asynchronous. If a command is synchronous or may be either synchronous or asynchronous depending on how it is invoked, you need to look at other details about that command.

`Start` is also an appropriate verb when you are starting a longer running process, even it is not stoppable/async and you block until that process completes. e.g. `Start-Sleep`, or for contributors to PowerShell core, `Start-PSBuild`, `Start-PSBootstrap`, or `Start-PSPester`.

`Invoke` is an appropriate verb when you are invoking an API or endpoint or scriptblock or PowerShell expression or some other dev-level command. e.g. `Invoke-RestMethod`, `Invoke-WebRequest`, `Invoke-WmiMethod`, `Invoke-Command`, `Invoke-History`, `Invoke-Expression`, `Invoke-AzureRmOperationalInsightsQuery`.

`Invoke` is also appropriate when you do not know the outcome of the command, or when multiple verbs could apply depending on what is being invoked. e.g. `Invoke-Item`. Many of the examples in the previous paragraph apply here as well, because the outcome of those commands is determined only when the command is invoked.

Lastly, `Invoke` is appropriate when you want to maintain a relationship with a pre-existing domain-specific vocabulary, and none of the other PowerShell verbs line up properly to that vocabulary. e.g. `Invoke-DbaAgFailover`, `Invoke-AzureRmOperationalInsightsQuery`.

> To be clear: I think what your RFCs propose would make for useful additions to the language, but they are unrelated to our conversation here.

They are absolutely related to our conversation. These RFCs, along with others, are currently ongoing discussions that will eventually lead to more asynchronous options for command authors (and scripters). As they mature and become implemented features in the language, I expect commands to evolve, such that they may support both synchronous and asynchronous processing, or such that they may change from synchronous processing to asynchronous processing. Such changes should not result in commands being renamed just because their internal processing has changed.

> > We just need to update the docs that describe when you would use these verbs. It's that simple.
> 
> The docs already do that, and, to my mind, sufficiently.

I feel this is obviously not the case, because if the docs were sufficient you wouldn't have proposed we rename `Start-Sleep` and we wouldn't be having this discussion about docs calling out something that doesn't need to be called out. I don't blame you for proposing such changes -- you were doing so based on how these verbs were documented in the guidelines; however, I stand by the point that the guidelines are documented incorrectly when it comes to when to choose `Start` and when to choose `Invoke`.

IMHO each of the points I made above should be made clear in the verb guidelines.

I would replace this:

> ### Invoke vs. Start
> 
> The `Invoke` verb is used to perform an operation that is generally a synchronous operation, such as running a command. The `Start` verb is used to begin an operation that is generally an asynchronous operation, such as starting a process.

With this:

> ### Start vs. Invoke
> 
> If your command always runs asynchronously, use `Start` (e.g. `Start-Job`, `Start-Process`).
>
> If your command may take longer to run, use `Start` (e.g. `Start-PSPester`, `Start-Sleep`).
>
> If your command invokes an API, endpoint, or other dev-level command, use `Invoke` (e.g. `Invoke-Command`, `Invoke-RestMethod`).
>
> If the outcome of your command is only known when the command is invoked, use `Invoke`  (e.g. `Invoke-Item`).
>
> If it is important for your command to maintain a relationship with a pre-existing domain-specific vocabulary (DSV), and none of the other PowerShell verbs line up properly to that vocabulary, use `Invoke` (e.g. `Invoke-DbaAgFailover`). Note that in this case, aliases can be used to maintain relationships with DSVs, allowing you to use approved verbs that are appropriate but perhaps are not a direct match to your DSV while still facilitating discovery of your commands.

**Comment by mklement0** [comment]
> Start is also an appropriate verb when you are starting a longer running process, even it is not stoppable/async and you block until that process completes

It is this very _conflation of two unrelated aspects_ that robs the verb `Start` of its specificity:

_How long something takes_ is independent of whether _something is synchronous or asynchronous_, and by treating the two the same, a `Start-`command no longer conveys specific information, leaving aside the fact that "longer running" is inherently vague.

`Start-Sleep` is a particularly bad example, because it is neither: it isn't asynchronous, and it doesn't take a long time: it takes exactly _how long you tell it to take_ (and that's typically a _short_ time).
(The other misnamed commands fortunately have a limited audience, due to only being used by those building PowerShell themselves.)

> that will eventually lead to more asynchronous options for command authors (and scripters)

Introducing generic commands / operators that facilitate various ways of providing asynchronous operation of _other, intrinsically synchronous commands_ is great, and, as stated, a worthwhile pursuit.

That you can use an intrinsically-or-by-default-synchronous command with other commands / operators that situationally make it asynchronous has nothing to do with what that command should be named.

> Such changes should not result in commands being renamed just because their internal processing has changed.

A command's (default) synchrony is  a fundamental aspect of its behavior, and very much part of its public contract - not internal processing.

Since such fundamental aspects shouldn't be changed, the question of a name change is moot.
If such a breaking change were to be considered, then the name _should_ be changed to reflect it, or a command variant under a new, appropriate name should be introduced.

**Comment by KirkMunro** [comment]
> It is this very conflation of two unrelated aspects that robs the verb Start of its specificity.

No, it's not.

`Start` vs `Invoke` wasn't documented as exclusively asynchronous vs synchronous in the first place, because it has never been about that. This is exactly why the documentation includes "generally" in the description. `Start` -ne "asynchronous". Not now, and not since v1. That very problem, and the fact that you want to erase the "generally" aspect of the documentation, is how we got here.

Consider: `Invoke-Item` vs. `Start-Process`.

You're saying you want `Start` to be very specifically (and only) asynchronous. It has never been that clear cut, hence the word "generally" in the documentation.

I've tried suggesting why that is the case, with a proposal for additional documentation describing how the two verbs could be used, but you seem stuck on making `Start` -eq "asynchronous".

🤷‍♂

**Comment by mklement0** [comment]
We've gotten to the heart of the issue:

* You think that it's acceptable for a given approved verb to express multiple, unrelated meanings.

* I think that allowing that undermines the very purpose of maintaining a list of approved verbs.

I think it is problematic to interpret "generally" as "could also mean something unrelated".

`Invoke-Item` is the only case that to my mind exemplifies a useful reading of "generally", i.e, a legitimate exception:

It is inherently open-ended whether `Invoke-Item` is synchronous or not, because the behavior is _up to each implementing provider_. In the only standard provider that implements it - the file-system provider - it _happens to be_ asynchronous.

`Start-` is asynchronous in all standard cmdlets except `Start-Sleep`, and documenting that aberration is all that is needed, perhaps complemented with a note  in the `Invoke-Item` topic.

(As stated, few people will see the commands in `build.psm1`; speaking of: given that `Build` is an approved verb, something like `Start-PSBuild` is especially baffling.)

I appreciate your detailed proposal to amend the docs, but I think it's the wrong way to go.

Debating this as much as we have may seem pedantic, but I think it's important to have clarity on such fundamental matters.
